runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#ifndef RUNTIME_MM_OBJECT_FACTORY_H
+#define RUNTIME_MM_OBJECT_FACTORY_H
+
+#include <algorithm>
+#include <memory>
+#include <mutex>
+
+#include "Alignment.hpp"
+#include "Alloc.h"
+#include "Memory.h"
+#include "Mutex.hpp"
+#include "Utils.hpp"
+
+namespace kotlin {
+namespace mm {
+
+namespace internal {
+
+// A queue that is constructed by collecting subqueues from several `Producer`s.
+// This is essentially a heterogeneous `MultiSourceQueue` on top of a singly linked list that
+// uses `konanAllocMemory` and `konanFreeMemory`
+// TODO: Consider merging with `MultiSourceQueue` somehow.
+class ObjectFactoryStorage : private Pinned {
+public:
+    // This class does not know its size at compile-time.
+    class Node : private Pinned {
+    public:
+        ~Node() = default;
+
+        static void operator delete(void* ptr) noexcept { konanFreeMemory(ptr); }
+
+        // Note: This can only be trivially destructible data, as nobody can invoke its destructor.
+        void* Data() noexcept { return data_; }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        explicit Node(void* data) noexcept : data_(data) {}
+
+        static std::unique_ptr<Node> Create(size_t dataSize, size_t dataAlignment) noexcept {
+            RuntimeAssert(IsValidAlignment(dataAlignment), "dataAlignment=%zu is not a valid alignment", dataAlignment);
+            RuntimeAssert(IsAligned(dataSize, dataAlignment), "dataSize=%zu must be aligned to dataAlignment=%zu", dataSize, dataAlignment);
+            size_t alignment = std::max(alignof(Node), dataAlignment);
+            size_t size = sizeof(Node) + dataSize;
+            size_t allocSize = AlignUp(size, alignment);
+            void* ptr = konanAllocMemory(allocSize);
+            if (!ptr) {
+                // TODO: Try doing GC first.
+                konan::consoleErrorf("Out of memory trying to allocate %zu. Aborting.\n", allocSize);
+                konan::abort();
+            }
+            void* alignedPtr = AlignUp(ptr, alignment);
+            RuntimeAssert(
+                    reinterpret_cast<uintptr_t>(alignedPtr) + size <= reinterpret_cast<uintptr_t>(ptr) + allocSize,
+                    "Aligning %p (with size %zu) to %p overflowed allocated size %zu", ptr, size, alignedPtr, allocSize);
+            size_t dataOffset = AlignUp(sizeof(Node), dataAlignment);
+            void* data = static_cast<uint8_t*>(ptr) + dataOffset;
+            RuntimeAssert(IsAligned(data, dataAlignment), "data=%p is not aligned to %zu", data, dataAlignment);
+            auto* nodePtr = new (ptr) Node(data);
+            return std::unique_ptr<Node>(nodePtr);
```

Comment:
The entire implementation above looks kinda cryptic, but doesn't look correct to me.
For example:
1. It seems to rely on the allocator providing pointers aligned by `alignment`, which might be not true.
2. Consider the case dataSize = 112, dataAlignment = 128. IIUC, the implementation would allocate 128 bytes with `data` pointing to the end of the region, and no assertion would fire (if the allocator returns memory aligned by 128).

It makes some sense to make the implementation simpler by specializing it to dataAlignment = 8 for now. For example, in this case node size and alignment would be aligned by dataAlignment.

Reply:
Oh, that's another mistake. Of course it should be `alignedPtr` instead of `ptr` in the call to `new`.

Reply:
> Consider the case dataSize = 112, dataAlignment = 128

This'll fail at the second assertion: `dataSize` must itself be aligned by `dataAlignment`.

Reply:
Reworked again and made `dataAlignment` a compile-time parameter.